frame/election-provider-multi-phase/src/unsigned.rs

```diff
@@ -60,33 +73,134 @@ impl From<FeasibilityError> for MinerError {
 	}
 }
 
+/// Save a given call into OCW storage.
+fn save_solution<T: Config>(call: &Call<T>) -> Result<(), MinerError> {
+	let storage = StorageValueRef::persistent(&OFFCHAIN_CACHED_CALL);
+	match storage.mutate::<_, (), _>(|_| Ok(call.clone())) {
+		Ok(Ok(_)) => Ok(()),
+		Ok(Err(_)) => Err(MinerError::FailedToStoreSolution),
+		Err(_) => unreachable!("mutation function cannot fail; qed"),
+	}
+}
+
+/// Get a saved solution from OCW storage if it exists.
+fn restore_solution<T: Config>() -> Result<Call<T>, MinerError> {
+	StorageValueRef::persistent(&OFFCHAIN_CACHED_CALL)
+		.get()
+		.flatten()
+		.ok_or(MinerError::NoStoredSolution)
+}
+
+/// Clear a saved solution from OCW storage.
+pub(super) fn kill_solution<T: Config>() {
+	let mut storage = StorageValueRef::persistent(&OFFCHAIN_CACHED_CALL);
+	storage.clear();
+}
+
 impl<T: Config> Pallet<T> {
-	/// Mine a new solution, and submit it back to the chain as an unsigned transaction.
-	pub fn mine_check_and_submit() -> Result<(), MinerError> {
+	/// Attempt to restore a solution from cache. Otherwise, compute it fresh. Either way, submit
+	/// if our call's score is greater than that of the cached solution.
+	pub fn restore_or_compute_then_maybe_submit() -> Result<(), MinerError> {
+		log!(
+			info,
+			"OCW attempting to restore or compute an unsigned solution for the current election"
+		);
+
+		let (call, score) = restore_solution::<T>()
+			.and_then(|call| {
+				// ensure the cached call is still current before submitting
+				let current_round = Self::round();
+				match &call {
+					Call::submit_unsigned(solution, _) if solution.round == current_round => {
+						let score = solution.score.clone();
+
+						log!(
+							debug,
+							"restored a cached call for round {} with score {:?}",
+							current_round,
+							score,
+						);
+
+						Ok((call, score))
+					}
+					_ => Err(MinerError::SolutionOutOfDate),
+				}
```

Comment:
```suggestion
				if let Call::submit_unsigned(solution, _) = call {
				    if solution.round == current_round { ... } else { MinerError::SolutionOutOfDate } 
				} else { 
				      MinerError::SolutionCallInvalid
				}
```

my main concern was that if `Call` is not a match we raise `SolutionOutOfDate` which is a bit misleading, but you could say if the call is not a match, then it can only happen in a runtime upgrade that changes the `solution` struct, and the solution is.... kinda outdated if you think about it like that. 

anyway, just a suggestion.

Reply:
https://github.com/paritytech/substrate/pull/8290/commits/2b5d0505aca55e73e0cff3bd0c9e4770aa122303